### DISCUSSION TITLE
[question]Getting a list of all strings

### DISCUSSION THREAD

**Discussion by KakCAT** [icsharpcode/ILSpy#2190] [Help] [answered]
Hi! 

I want to extract a list of all the strings in the assembly I'm building. Right now I have a function which, in runtime, detects when a string is printed to screen and logged into a DB for translation, but this has obvious flaws.

I just wanted to ask if ILSpy has this string extraction feature planned, or it's not even on the map.

Thanks!

**Comment by compasslg** [comment, marked as answer]
I think there is no need to use ILSpy for that, just Mono.Cecil is enough to do the job. Scan through all the IL instructions and look for the ldstr opcode.

**Comment by KakCAT** [comment]
Thanks, I'll try that, it sounds really easy to do.

**Reply by siegfriedpammer** [reply]
If you only want to get the UserStrings, you should not use Mono.Cecil, but rather System.Reflection.Metadata. Cecil is very heavy-weight for such a task, because it creates an object for each instruction, however IL is very easy to parse just by looking at the bytes. There are some extensions in https://github.com/icsharpcode/ILSpy/blob/master/ICSharpCode.Decompiler/Disassembler/ILParser.cs you can use. Just looking at the bytes will have a very big performance/memory foot print advantage.

**Reply by siegfriedpammer** [reply]
If you need examples, how to work with SRM and parse bytes you can take a look at one of the many analyzers in ILSpy. https://github.com/icsharpcode/ILSpy/blob/master/ILSpy/Analyzers/Builtin/MethodUsesAnalyzer.cs#L70-L138 for example. ILSpy uses SRM + a few extension methods, which you could easily copy.

**Comment by KakCAT** [comment]
Hi @siegfriedpammer , thanks for your message.

I went with the Mono.Cecil because it seemed easier (actually it was incredibly easy and fast to do,much more than I expected). After all, this process will be run only once a month (or even less often) over a single assembly.

However if I ever need to do something more complex that needs more performance, I'll try your method.

Thanks a lot for the answers.
